proposals/0064-transaction-receipt.md

```diff
@@ -0,0 +1,416 @@
+
+Nα
+|
+0
+```
+
+```txt
+Transaction receipt tree with four transaction receipts as leaf
+nodes [L0, L1, L2, L3] where R0, R1, R2, R3 are the transaction 
+receipts and Nδ is the root.
+           Nδ
+          /  \
+        /     \
+       Nγ      N(transaction_receipts)
+      /  \
+     /    \
+   Nα      Nβ
+  / |     / \
+ /  |    /   \
+L0  L1   L2  L3
+
+L0 := sha256(concat(0x00, R0))
+L1 := sha256(concat(0x00, R1))
+L2 := sha256(concat(0x00, R2))
+L3 := sha256(concat(0x00, R3))
+Nα := sha256(concat(0x01, hash(L0), hash(L1)))
+Nβ := sha256(concat(0x01, hash(L2), hash(L3)))
+Nγ := sha256(concat(0x01, hash(Nα), hash(Nβ)))
+Nδ := sha256(concat(0x80, hash(Nγ),len([L0, L1, L2, L3])))
+```
+
+```txt
+Transaction receipt tree with five transaction receipts 
+as leaf nodes [L0, L1, L2, L3, L4] where R0, R1, R2, R3, R4 are the 
+transaction receipts and Nτ is the root.
+                Nτ   
+              /   \
+            /      \
+          Nζ        N(transaction_receipts)
+         /  \
+        /    \
+       Nδ     Iε
+      /  \     \\
+     /    \     \\
+   Nα      Nβ    Nγ
+  /  \    /  \   ||
+ L0  L1  L2  L3  L4
+
+L0 := sha256(concat(0x00, R0))
+L1 := sha256(concat(0x00, R1))
+L2 := sha256(concat(0x00, R2))
+L3 := sha256(concat(0x00, R3))
+L4 := sha256(concat(0x00, R4))
+Nα := sha256(concat(0x01, hash(L0), hash(L1)))
+Nβ := sha256(concat(0x01, hash(L2), hash(L3)))
+Nγ := sha256(concat(0x01, hash(L4), hash(L4)))
+Nδ := sha256(concat(0x01, hash(Nα), hash(Nβ)))
+Nζ := sha256(concat(0x01, hash(Nδ), hash(Iε)))
+Nτ := sha256(concat(0x80, hash(Nζ), len([L0, L1, L2, L3, L4])))
+
+Here 'Nτ' is the root generated after concatenating
+the node 'Nζ' with the length of vector of leaf nodes (which is five in the above
+illustration) of the tree and hashing it.
+```
+
+#### Benchmarks
+
+Tinydancer has prepared benchmarks comparing two merkle tree implementations.
+The input used was a vector of 500k TransactionReceiptData objects.
+
+1) Solana Labs Merkle Tree: This is the pure Rust implementation that is currently
+   used by the Solana Labs client.
+   [Solana Labs GitHub](https://github.com/solana-labs/solana/tree/master/merkle-tree)
+2) Firedancer binary merkle tree (bmtree): Implemented in C and uses Firedancer's
+   optimised SHA-256 implementation. Benchmarks were performed via Rust FFI bindings,
+   and thus may not be representative of the upper bound.
+   An improved version using AVX-512 instructions is in progress.
+   [Firedancer GitHub](https://github.com/firedancer-io/firedancer/tree/main/src/ballet/bmtree)
+
+[Benchmark Source](https://github.com/tinydancer-io/merkle-bench)
+![TINY_bench](https://github.com/tinydancer-io/solana-improvement-documents/assets/50767810/637dc83a-b3d2-4616-b70e-4fbb8a9e17fd)
+
+## Impact
+
+This would enable SIMD-0052 to be implemented, which adds the transaction receipt
+root to the bank hash. This will allow users to verify transaction statuses returned
+by third-party RPC nodes.
```

Comment:
```suggestion
The generation of a Transaction Receipt Tree is a prerequisite for providing proof
that a transaction was included in a block. Itself a step toward providing proof
that a transaction was executed and accepted under consensus by a Solana cluster. A
major improvement in trust-minimization for the ecosystem, opening the door to new
use-cases and reducing infrastructure requirements for some of today's.
```

this only gives us inclusion of a transaction into a block, not a block under consensus of the majority fork. a step in the right direction, but insufficient to verify an rpc response

Reply:
Ohk this is a fair detail to add.